### DISCUSSION TITLE
Google trying to index the literal dynamic route url

### DISCUSSION THREAD

**Discussion by ugurnarin** [vercel/next.js#37833] [Help] [answered]
Hi,

I have a Next project deployed on vercel, and got a warning (Server Error) this morning about google trying to index a dynamic route url. 

Specifically this one below, which is not in my sitemap or robots file and neither linked by mistake anywhere in the project.

domain-name.com/rhd-vehicles/[id]

I am confused as to how to fix this one, I guess I could disallow it on robots but now worried it might effect the pre-generated pages through this dynamic page setup. (Using getStaticProps) Since, google shouldn't even see it in the first place, and may tie it to the actual links pre-generated. 

My only guess is, it is embedded somewhere in the compiled code and google is following it. 

Did anyone experience something like this? Would appreciate any feedback. Project uses version "12.1.6"

Thanks.

**Comment by ugurnarin** [comment]
![server error](https://user-images.githubusercontent.com/79957810/174506275-7371616e-2f0e-408b-8895-82be857d5f3d.JPG)

**Comment by icyJoseph** [comment]
I've seen this problem before, do you have a repository to look at? The previous time I saw this the person didn't have a repo to share, and we couldn't really solve it.

My guess with this one is that there's a link pointing to `/rhd-vehicles/[id]` in the page's source, or that perhaps the bot is following client side redirects, but without source code it is hard to tell.

**Comment by ugurnarin** [comment]
Hi Joseph, 

Thank you for your reply, your help is appreciated. I don't have the repo available to the public, but I found where it is coming from and will share the code below. 

After what you mentioned, I looked at the actual page source of the website (I was just looking at the source code), inside the pages pre-generated with this dynamic route, there is a "page" value passed in the data source. See `"page":` below. P.S I trimmed the unrelated data/id.

```html
<script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"loadedVehicle":{JUST-SOME-DATA-HERE},"__N_SSG":true},"page":"/rhd-vehicles/[id]","query":{"id":"JUST-SOME-ID"},"buildId":"Tt2NspVuXHtZBz9zlZGMu","isFallback":false,"gsp":true,"scriptLoader":[]}</script></body></html>
```

Redirects only point to hard-coded URLs and there are only a few pointing to static pages, so it is not coming from there.

This has to be automatically generated by NextJS, because I don't pass this in the `getStaticProps` and have no idea what could be triggering this.  

This is the server-side code used in this same route.

```js
async function getData() {
	const filePath = path.join(process.cwd(), 'data', 'data.json');
	const jsonData = await fs.readFile(filePath);
	const data = JSON.parse(jsonData);

	return data;
}

export async function getStaticProps(context) {
	const url = context.params.id.split('-');
	const id = url[url.length - 1];

	const data = await getData();
	const vehicle = data.find((item) => item.id === Number(id));

	const vehicleData = {
		JUST-SOME-DATA
	};

	return {
		props: {
			loadedVehicle: vehicleData,
			url: url,
		},
	};
}

export async function getStaticPaths() {
	const data = await getData();

	const pathsWithParams = data.map((item) => ({
		params: {
			id: `/rhd-vehicles/${item.year}-${item.brand
				.trim()
				.replaceAll(' ', '-')}-${item.model
				.trim()
				.replaceAll(' ', '-')}-${item.grade
				.trim()
				.replaceAll(' ', '-')}-${item.id}`,
		},
	}));

	return {
		paths: pathsWithParams,
		fallback: 'blocking',
	};
}
```

**Comment by icyJoseph** [comment, marked as answer]
Alright, looks like this is the issue then, https://github.com/vercel/next.js/discussions/35747.

According to, https://stackoverflow.com/questions/47210596/how-to-prevent-google-from-indexing-script-type-application-json-content, the bot parses the `application/json` scripts, and though we can tell the bot to not follow them, the bot is still crawling the content.

As far as I know there's no control mechanism over this, other than explicitly telling the bot to not crawl those URLs (containing `[]`), but that's about it.

**Comment by ugurnarin** [comment]
Thanks Joseph, I will just add it to the robots file.

**Reply by danielbeckera** [reply]
Hello, this worked?
Can you show me how you've added this condition on robots file?

**Reply by ugurnarin** [reply]
"Disallow: /rhd-vehicles/[id]" 

I also have a sitemap that covers every generated page link from this.

**Comment by mcnaveen** [comment]
Hi @leerob Does Nextjs docs cover this? Am I missing something? 🤔

**Reply by mcnaveen** [reply]
@leerob
